File: test/nu.c
#include <stdio.h>
#include <stdbool.h>
#include <stdint.h>

// testing void as root 

typedef void nu;

enum {
    nu_error_none = 0,
    nu_error_generic,
    nu_error_file,
    nu_error_invalid_arg,
};

nu* nu_get() {
    return (nu*) 0x1000;
}

int nu_hello(nu* handle) {
    return nu_error_none;
}

#define nu_option_one     0x0001
#define nu_option_two     0x0002
#define nu_option_three     0x0004

typedef struct { 
    int number;
    uint16_t options;
} nu_struct;

int nu_world(nu* handle, uint16_t options) {
    if (options & nu_option_one) {
        return nu_error_none;
    } else if (options & nu_option_two) {
        return nu_error_generic;
    }
}

void nu_foo(nu* handle, uint16_t options, nu_struct* ptr) {
    ptr->number = 1234;
    ptr->options = options;
}

File: test/nu.zig
const std = @import("std");
const api_translator = @import("api-translator.zig");
const c = @cImport({
    @cInclude("nu.c");
});

pub const Error = error{
    Generic,
    File,
    InvalidArg,
    Unexpected,
    NullPointer,
};
pub const ErrorEnum = enum(c_int) {
    none,
    generic,
    file,
    invalid_arg,
    _,
};

pub const get: fn () Error!*@This() = c_to_zig.translate("nu_get", true, false, .{});

pub const hello: fn (
    handle: *@This(),
) Error!void = c_to_zig.translate("nu_hello", true, false, .{});

pub const Struct = extern struct {
    number: c_int,
    options: Options,
};

pub const world: fn (
    handle: *@This(),
    options: Options,
) Error!void = c_to_zig.translate("nu_world", true, false, .{});

pub const foo: fn (
    handle: *@This(),
    options: Options,
) Struct = c_to_zig.translate("nu_foo", false, false, .{});

pub const Options = packed struct(u16) {
    one: bool = false,
    two: bool = false,
    three: bool = false,
    _: std.meta.Int(.unsigned, @bitSizeOf(u16) - 3) = 0,
};
const c_to_zig = api_translator.Translator(.{
    .c_import_ns = c,
    .substitutions = &.{
        .{ .old = ?*c.nu, .new = *@This() },
        .{ .old = c.nu_struct, .new = Struct },
        .{ .old = u16, .new = Options },
    },
    .error_scheme = api_translator.BasicErrorScheme(ErrorEnum, Error, Error.Unexpected, .{
        .{ .type = *@This(), .err_value = null, .err = error.NullPointer },
    }),
});

test {
    inline for (comptime std.meta.declarations(@This())) |decl| {
        _ = @field(@This(), decl.name);
    }
}
